com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/util/BobProjectProperties.java

```diff
@@ -25,34 +25,299 @@
+            String[] values = this.value.trim().split(",");
+            Map<Integer, String> outValues = new TreeMap<Integer, String>();
+            int counter = 0;
+            for (String v : values) {
+                v = v.trim();
+                if (v.length() > 0) {
+                    outValues.put(counter, v);
+                    counter++;
+                }
+            }
+            this.valuesArray = outValues;
+        }
+
+    }
+
+    private Map<String, Map<String, ProjectProperty>> properties;
 
     /**
      * Constructor with initially empty
      */
     public BobProjectProperties() {
-        this.properties = new LinkedHashMap<String, Map<String, String>>();
-        this.defaults = new LinkedHashMap<String, Map<String, String>>();
+        this.properties = new LinkedHashMap<String, Map<String, ProjectProperty>>();
+    }
+
+    public Map<String, Map<String, Object>> createTypedMap(PropertyType[] types) {
+        Map<String, Map<String, Object>> result = new HashMap<String, Map<String, Object>>();
+        for (String categoryName : properties.keySet()) {
+            Map<String, Object> resultCategory = new HashMap<String, Object>();
+            result.put(categoryName, resultCategory);
+            Map<String, ProjectProperty> category = properties.get(categoryName);
+            for (String key : category.keySet()) {
+                ProjectProperty prop = category.get(key);
+                if (ArrayUtils.contains(types, prop.getType())) {
+                    resultCategory.put(key, prop.getTypedValue());
+                }
+                else {
+                    resultCategory.put(key, prop.getValue());
+                }
+                
+            }
+        }
+        return result;
+    }
+
+    /**
+     * Load properties in-place from {@link InputStream}
+     * @param in {@link InputStream} to load from
+     * @param isMeta if loads meta file or not
+     * @throws IOException
+     * @throws ParseException
+     */
+    public void load(InputStream in, boolean isMeta) throws IOException, ParseException {
+        try {
+            Map<String, Map<String, ProjectProperty>> props = doLoad(in, isMeta);
+            // merge into properties
+            BobProjectProperties.mergeProperties(properties, props);
+        } finally {
+            IOUtils.closeQuietly(in);
+        }
+    }
+
+    /**
+     * Load properties in-place from {@link InputStream} as non-meta file
+     * @param in {@link InputStream} to load from
+     * @throws IOException
+     * @throws ParseException
+     */
+    public void load(InputStream in) throws IOException, ParseException {
+        load(in, false);
+    }
+
+    /**
+     * Load default properties from the internal meta.properties file
+     * @throws IOException
+     * @throws ParseException
+     */
+    public void loadDefaultMetaFile()  throws IOException, ParseException {
+        InputStream is = Bob.class.getResourceAsStream("meta.properties");
```

Comment:
Should this name also have a variable?
E:g. like `PROPERTIES_EXTENSION_FILE` does.